File: agri_coldchain/agri_cold_chain/report/transit_breach_log/transit_breach_log.py
from __future__ import unicode_literals

import frappe
from frappe import _


def execute(filters=None):
    """Transit Breach Log — Query Report.

    Filters Delivery Transit Log records where temperature_breach = 1.
    """
    columns = [
        {"fieldname": "name", "label": _("Transit Log ID"), "fieldtype": "Link", "options": "Delivery Transit Log", "width": 180},
        {"fieldname": "delivery_note", "label": _("Delivery Note"), "fieldtype": "Link", "options": "Delivery Note", "width": 180},
        {"fieldname": "transporter_name", "label": _("Transporter"), "fieldtype": "Data", "width": 150},
        {"fieldname": "vehicle_no", "label": _("Vehicle No"), "fieldtype": "Data", "width": 120},
        {"fieldname": "dispatch_temp", "label": _("Dispatch Temp (°C)"), "fieldtype": "Float", "width": 120},
        {"fieldname": "arrival_temp", "label": _("Arrival Temp (°C)"), "fieldtype": "Float", "width": 120},
        {"fieldname": "creation", "label": _("Log Date"), "fieldtype": "Date", "width": 100},
    ]

    conditions = "WHERE dtl.temperature_breach = 1"
    if filters:
        if filters.get("from_date"):
            conditions += " AND dtl.creation >= %(from_date)s"
        if filters.get("to_date"):
            conditions += " AND dtl.creation <= %(to_date)s"
        if filters.get("transporter_name"):
            conditions += " AND dtl.transporter_name LIKE %(transporter_name)s"

    sql = f"""
        SELECT
            dtl.name,
            dtl.delivery_note,
            dtl.transporter_name,
            dtl.vehicle_no,
            dtl.dispatch_temp,
            dtl.arrival_temp,
            DATE(dtl.creation) AS creation
        FROM `tabDelivery Transit Log` dtl
        {conditions}
        ORDER BY dtl.creation DESC
    """

    data = frappe.db.sql(sql, filters or {}, as_dict=True)

    return columns, data

File: agri_coldchain/agri_cold_chain/report/transit_breach_log/transit_breach_log.js
// Copyright (c) 2025, Your Organisation and contributors
// For license information, please see license.txt

frappe.query_reports["Transit Breach Log"] = {
    filters: [
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.add_months(frappe.datetime.get_today(), -3),
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
        },
        {
            fieldname: "transporter_name",
            label: __("Transporter"),
            fieldtype: "Data",
        },
    ],
    formatter: function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "arrival_temp" && data) {
            if (data.arrival_temp > 10) {
                value = `<span style="color:#e74c3c;font-weight:bold;">${value}°C 🚨</span>`;
            } else if (data.arrival_temp > 4) {
                value = `<span style="color:#e67e22;font-weight:bold;">${value}°C ⚠</span>`;
            }
        }

        return value;
    },
};
